spyql/query_result.py

```diff
@@ -0,0 +1,62 @@
+from typing import Any
+
+
+class QueryResult(tuple):
+    """
+    Result of a query that writes outputs to memory.
+    Tuple of dictionaries with easy access of columns by name as attributes.
+
+    Accessing the value of the `age` column in the first row::
+
+        result[0].age
+        result[0]["age"]
+
+    Collecting the age for all rows as a tuple::
+
+        result.age
+        result.col("age")
+
+    Collecting the age for a subset of rows as a tuple::
+
+        result[1:3].age
+        result[1:3].col("age")
+
+    Collecting the value of the first column for all rows as a tuple::
+
+        result.col(0)
+
+    Iterating over rows::
+
+        for row in result:
+            print(row.age, row.another_column)
+
+    """
+
+    def __new__(cls, __values, __colnames):
+        return super(QueryResult, cls).__new__(cls, __values)
+
+    def __init__(self, __values, __colnames):
+        self.__colnames = tuple(__colnames)
+
+    def col(self, idx):
+        """
+        Collects and returns a tuple with all values of the col defined by `idx`.
+        If `idx` is an integer it refers to the nth column (0-based indexing).
+        If `idx` is a string it refers to the name of the column.
+        """
+        if isinstance(idx, int):
+            idx = self.__colnames[idx]
```

Comment:
❓ Shouldn't we return here? Otherwise `idx` will be assigned, but it will return a type error.

Reply:
Since we are storing a tuple of dicts, we can only access columns by column name (dict key).
So, what we are doing here is, when the user wants to access by column number, we first get the column name (and replace the value of `idx` by the column name) and then run the rest of the code as if the user passed the column name as argument.